[DTP-963] Add support for customer provided typings for LiveObjects

VeskeR opened the pull request:
This PR is based on https://github.com/ably/ably-js/pull/1921, please review it first.

Users can provide their own typings for LiveObjects structure by declaring global `LiveObjectsTypes` interface:

```typescript
// ably.config.d.ts - example file name where to define the interface
declare global {
  export interface LiveObjectsTypes {
    root: {
      numberKey: number;
      stringKey: string;
      booleanKey: boolean;
      couldBeUndefined?: string;
      mapKey?: LiveMap<{
        foo: 'bar';
        nestedMap?: LiveMap<{
          baz: 'qux';
        }>;
      }>;
      counterKey?: LiveCounter;
    };
  }
}
```

or alternatively by providing the state tree structure for the `liveObjects.getRoot` call (useful if they user multiple channels with different liveobjects structure):
```typescript
const root = await liveObjects.getRoot<{ test: number; anotherCounter: LiveCounter }>();
root.get('test'); // number
root.get('anotherCounter'); // LiveCounter
```

Resolves [DTP-963](https://ably.atlassian.net/browse/DTP-963)

[DTP-963]: https://ably.atlassian.net/browse/DTP-963?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Enhanced TypeScript definitions for the Ably Realtime and REST client library, improving type safety and flexibility.
	- Introduced a new global interface `LiveObjectsTypes` for customizable LiveObjects structures.
	- Added a new type `ExplicitRootType` for better root object structure definition.

- **Bug Fixes**
	- Improved type checks and handling of LiveObjects in the `testAblyPackage` function.

- **Documentation**
	- Added comments and documentation for clarity on new types and methods.
	- Configured documentation settings to exclude certain internal types from public API documentation.

- **Chores**
	- Updated TypeScript compiler options to enforce stricter null checks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mschristensen reviewed (commented):
Looks great, thanks Andrii! Couple of suggestions

VeskeR commented on test/package/browser/template/src/index-liveobjects.ts:
```diff
@@ -13,9 +15,23 @@ globalThis.testAblyPackage = async function () {
+  const baz: 'qux' = map?.get('nestedMap')?.get('baz')!;
+
   // check LiveMap subscription callback has correct TypeScript types
   const { unsubscribe } = root.subscribe(({ update }) => {
```
created https://ably.atlassian.net/browse/DTP-1094 for this

VeskeR commented on test/package/browser/template/src/index-liveobjects.ts:
```diff
@@ -29,10 +45,8 @@ globalThis.testAblyPackage = async function () {
-  // so the next calls would fail. we only need to check that TypeScript types work
+  // check LiveCounter type also behaves as expected
+  // same deal with nullish coalescing
   counter?.value();
```
good idea, will add

VeskeR replied:
added assigning to a typed variable

VeskeR commented on test/package/browser/template/src/index-liveobjects.ts:
```diff
@@ -13,9 +15,23 @@ globalThis.testAblyPackage = async function () {
   const root = await liveObjects.getRoot();
```
true, will add

VeskeR replied:
added in https://github.com/ably/ably-js/compare/f921ce9147a0ad109c040343d0ef6c7838108bec..a1e29ff78a1140d332480a003dd547c5a32ef976

VeskeR commented on ably.d.ts:
```diff
@@ -2029,13 +2029,32 @@ export declare interface PushChannel {
+  getRoot<T extends LiveMapType = DefaultRoot>(): Promise<LiveMap<T>>;
 }
 
 // TODO: document public API for LiveObjects
```
yeah, initiall I thought I can postpone writing the docstrings until all the public API is in there (write API is missing rn), but we have the API reference test that expects all types in `ably.d.ts` to be documented, so I guess I'll do it now

VeskeR replied:
extracted read-only API documentation to a separate PR, see https://github.com/ably/ably-js/pull/1928, and updated this PR to also add documentation for new types, see https://github.com/ably/ably-js/compare/f921ce9147a0ad109c040343d0ef6c7838108bec..a1e29ff78a1140d332480a003dd547c5a32ef976.
TODO is removed

VeskeR commented on ably.d.ts:
```diff
@@ -2029,13 +2029,32 @@ export declare interface PushChannel {
+}
+
+// LiveMap type representation of how it looks to the user. A mapping of string keys to scalar values (StateValue) or other Live Objects.
+export type LiveMapType = { [key: string]: StateValue | LiveMap<LiveMapType> | LiveCounter | undefined };
```
I just used the same approach we already had in the public types for the `CipherKey` and `CipherKeyParams` https://github.com/ably/ably-js/blob/18a255948c38d1e60715c8f5d6173369b57cb8d6/ably.d.ts#L2527-L2531, which is  to include both types for node.js and browsers (though an additional comment in the docstring should be added to specify which type you will receive on which platform).

I think to make it platform dependant we would need to have two separate type declaration files, and configure their exports in `package.json` so node.js and browser environments get the appropriate one.

It is way beyond the scope of this PR and LiveObjects work, and probably has too little of an impact relative to the amount of work required to think about it right now

VeskeR replied:
updated comment for `StateValue` in https://github.com/ably/ably-js/compare/f921ce9147a0ad109c040343d0ef6c7838108bec..a1e29ff78a1140d332480a003dd547c5a32ef976

VeskeR commented on test/package/browser/template/src/index-liveobjects.ts:
```diff
@@ -2,6 +2,8 @@ import * as Ably from 'ably';
 import LiveObjects from 'ably/liveobjects';
 import { createSandboxAblyAPIKey } from './sandbox';
 
+// TODO: check we can refer to types exported by LiveObjects plugin.
```
updated comments and removed this TODO

VeskeR commented:
@mschristensen  Addressed your feedback, fixed package and API reference tests.
Package tests also now succeeds locally for me via `npm run test:package`

mschristensen reviewed (approved):
Looks great
